### PR TITLE
Observe using RxJava

### PR DESCRIPTION
**Pull request by sbosley**
This PR introduces a new addon, squidb-reactive. This module contains only a single class, ReactiveSquidDatabase, which is an extension of SquidDatabase allowing for RxJava style observation of table changes along the lines of SQLBrite. 

ReactiveSquidDatabase introduces six new methods:
* `observeTable(SqlTable<?> table)` returns an Observable that will emit the table itself each time the table is written to
* `observeTableAndEmit(SqlTable<?> table, T objectToEmit)` returns an Observable that will emit the given object each time the table is written to. You could for example pass a Query to be run.
* `observeTablesAndEmit(Collection<SqlTable<?>> tables, T objectToEmit)` returns an Observable that will emit the given object each time any of the given tables are written to
* An version of each of these three methods that also takes a boolean argument "emitOnFirstSubscribe"

Example usage:
```java
AtomicInteger callCount = new AtomicInteger();
Observable<AtomicInteger> observable = database.observeTableAndEmit(Person.TABLE, callCount);
Subscription s = observable.subscribe(new Action1<AtomicInteger>() {
    @Override
    public void call(AtomicInteger counter) {
        counter.incrementAndGet();
    }
});
// Observable will not get notifications on first subscribe unless the observable was created using
// the emitOnInitialSubscribe flag set to true
assertEquals(0, callCount.get());
database.beginTransaction();
try {
    database.persist(new Person().setFirstName("A").setLastName("B")
                    .setBirthday(System.currentTimeMillis()));
    database.persist(new Person().setFirstName("C").setLastName("D")
                    .setBirthday(System.currentTimeMillis()));
    database.persist(new Person().setFirstName("E").setLastName("F")
                    .setBirthday(System.currentTimeMillis()));
    database.setTransactionSuccessful();
} finally {
     database.endTransaction();
}
// Count only incremented once
assertEquals(1, callCount.get());

// Unsubscribe to stop getting notifications
s.unsubscribe();

database.persist(new Person().setFirstName("E").setLastName("F")
                    .setBirthday(System.currentTimeMillis()));
// Count hasn't changed
assertEquals(1, callCount.get());
```

This PR is against the dev_2.0 branch, so this change will not be on master until SquiDB 2.0 launches.

Refs #5

### CODE REVIEW

File: squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    /**
+     * @param objectToEmit an object for the created Observable to emit
+     * @param tables the tables to observe
+     * @return a new {@link Observable} that will be called whenever any of the given tables are written to in a
+     * successful statement or transaction. The Observable will emit the object passed in this method; for example, a
+     * Query to run.
+     */
+    public <T> Observable<T> observeTablesForObject(final Collection<? extends SqlTable<?>> tables, T objectToEmit) {
+        return observeForObject(objectToEmit, new Func1<Set<SqlTable<?>>, Boolean>() {
```
Comment (jdkoren) [inline squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java]: Do we need a null check on `tables`?

File: squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                }
+                return false;
+            }
+        });
+    }
+
+    private <T> Observable<T> observeForObject(final T objectToEmit, Func1<Set<SqlTable<?>>, Boolean> tableFilter) {
+        return changedTablePublisher
+                .filter(tableFilter)
+                .startWith(INITIAL_TABLE)
```
Comment (jdkoren) [inline squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java]: What's the purpose of this?

File: squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java (hunk 1 above)
Comment (sbosley) [inline squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java]: Probably a good idea, yeah.

File: squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java (hunk 2 above)
Comment (sbosley) [inline squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java]: It makes the first subscriber to this observable be immediately triggered. So for example, if you were emitting a Query you would immediately run the query and then re-run it on subsequent changes.

File: squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+    /**
+     * @param table the table to observe
+     * @param emitOnInitialSubscribe use true if you want the first subscriber to the observable to be triggered
+     * immediately; false otherwise
+     * @return a new {@link Observable} that will be called whenever the given table is written to in a successful
+     * statement or transaction. The Observable will emit the table itself; to emit some other object like a Query, use
+     * {@link #observeTableAndEmit(SqlTable, Object)}
+     */
+    public <T extends SqlTable<?>> Observable<T> observeTable(T table, boolean emitOnInitialSubscribe) {
```
Comment (jdkoren) [inline squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java]: minor nit: the others are named emitOnFirstSubscribe

Comment (sbosley) [inline squidb-addons/squidb-reactive/src/com/yahoo/squidb/reactive/ReactiveSquidDatabase.java]: Haha whoops.

**Comment by jdkoren** [comment]
LGTM. Travis is lookin' pretty green, I think I have just the right *prescription* to fix that. [@jdkoren dodges tomato]

**Comment by sdimmick** [comment]
@jdkoren :hear_no_evil::gun:

**Pull request merged**